### ISSUE TITLE
unpack-channel.nix fails in restricted eval mode

### ISSUE THREAD

**Issue by twhitehead** [NixOS/nix#2239]
Not entirely sure what to make of this, but figured I would open this issue to at least get some input.

I  built an overlay system to use on our super computer clusters in order to inject a serioes of  HPC related tweaks and packages into `nixpkgs` (I used to fork `nixpkgs`, but maintain  an separate overlay turned out to be a lot less work and works a lot better an automated hydra build process)

https://git.computecanada.ca/nix/ccpkgs

Since upgrading our hydra sever to 17.09 in mid-May, I am now getting an evaluation error

```
error: while evaluating the attribute 'distPhase' of the derivation 'ccpkgs-channel' at /nix/store/p9j197rlx2nhg5n15gp4kf03c88by3mb-git-export/pkgs/stdenv/generic/make-derivation.nix:98:11:
while evaluating the attribute 'chrootDeps' of the derivation 'nixos' at /nix/store/54jxw5h4r4ahl9mw3ydfpzxqmk07gdy6-nix-2.0/share/nix/corepkgs/unpack-channel.nix:35:10:
access to path '/nix/store/54jxw5h4r4ahl9mw3ydfpzxqmk07gdy6-nix-2.0' is forbidden in restricted mode
```

After  spending some time digging through things, it seems the  issue is that I use [`<nix/unpack-channel.nix>`](https://git.computecanada.ca/nix/ccpkgs/blob/a078e69b8c4314d90b9b5fbcbd8079e8c6388d1d/release.nix#L83) in the process of building to overlay channel, and _unpack-channel.nix_ does

```
with import <nix/config.nix>;
...
derivation {
  ...
  inherit chrootDeps;
}
```

where `chrootDeps` is defined in _nix/config.nix_ as

```
let
  ...
in rec {
  ...
  nixPrefix = "@prefix@";
  ...
  chrootDeps =
    if dirOf nixPrefix == builtins.storeDir then
      [ (builtins.storePath nixPrefix) ]
    else
      [ ];
}
```

Would welcome any input on this (i.e., whether it is something I should be working around, whether it is a corner case in nix that should be addressed, any general comments, etc.).

Thanks!  -Tyson

**Comment by edolstra** [comment]
Probably the best solution is to either

* Eliminate `<nix/unpack-channel.nix>` entirely by doing the unpacking in `nix-channel`. However, we would lose the implicit caching of unpacked channels.

* Make `<nix/unpack-channel.nix>` a builtin derivation (like `<nix/fetchurl.nix>`). However, it wouldn't be able to call `tar`, so we would have to have an in-process implementation of `tar x`.

**Comment by twhitehead** [comment]
Happened to notice today that you implemented the second option awhile back in 0a2bee3.

Tested it and I no longer get the restricted mode issue. Thanks! :smile:
